[#48114] Create Meetings outside of a Project

aaron-contreras opened the pull request:
See: https://community.openproject.org/work_packages/48114

aaron-contreras committed:
[REFACTOR] Prefix project-scoped routes with `project`

aaron-contreras committed:
Fix typo on Stimulus RefreshOnFormChangesController

dombesz commented on modules/meeting/config/routes.rb:
```diff
@@ -27,11 +27,11 @@
 OpenProject::Application.routes.draw do
-  scope 'projects/:project_id' do
-    resources :meetings, only: %i[new create index]
+  resources :projects, only: %i[] do
```
This might be a good candidate to use [shallow route nesting](https://guides.rubyonrails.org/routing.html#shallow-nesting).

dombesz commented on modules/meeting/app/controllers/projects/meetings_controller.rb:
```diff
@@ -0,0 +1,109 @@
+    @meetings = @project.meetings
+  end
+
+  def create
```
The authorization check for projects is missing from create. By altering the meeting form, I am able to create a meeting in a project where I don't have access to.
![Jun-15-2023 13-34-53](https://github.com/opf/openproject/assets/83396/0f750db8-9378-4762-8b3e-d917bc39c8fe)

dombesz reviewed (changes requested):
1. I'm not entirely sure if the two controllers are required, to simplify the code we could create a `meetings_scope` like:

```ruby
def meetings
  @projects = Project.allowed_to(User.current, :view_meetings)
  @projects = @projects.where(project_id:) if project_id   
  @meetings = Meeting.where(project: projects).from_today
end
```

This could be used in both of the index pages, without further ado.
The scope can be used to build the Meeting in the create action and it will automatically assign the `project_id`:

```ruby
def create
  @meeting = @meetings.build(meeting_params)
end
```

<img width="373" alt="image" src="https://github.com/opf/openproject/assets/83396/91a71538-1f5b-437e-88ae-2f162976efc0">

2. Submitting the meeting form without selecting a project results in an error page.

![Jun-15-2023 13-06-44](https://github.com/opf/openproject/assets/83396/6d228ed0-2d02-4a96-8fc7-096bf57c3fc8)

aaron-contreras committed:
Add Create Global Meetings Page

aaron-contreras commented on app/views/layouts/_common_head.html.erb:
```diff
@@ -6,7 +6,7 @@
 <meta name="app_base_path" content="<%= relative_url_root %>"/>
 <base href="<%= relative_url_root %>/" />
-<% if @project %>
+<% if !@global_create_context && @project %>
```
Thoughts on hiding the menu this way?

aaron-contreras commented on app/views/layouts/base.html.erb:
```diff
@@ -62,7 +62,7 @@ See COPYRIGHT and LICENSE files for more details.
 <op-modal-overlay></op-modal-overlay>
 <spot-drop-modal-portal></spot-drop-modal-portal>
-<% main_menu = render_main_menu(local_assigns.fetch(:menu_name, nil), @project) %>
+<% main_menu = render_main_menu(local_assigns.fetch(:menu_name, nil), !@global_create_context && @project) %>
```
Thoughts on hiding the menu this way?

aaron-contreras commented on modules/meeting/app/controllers/meetings_controller.rb:
```diff
@@ -28,15 +28,17 @@
 
 class MeetingsController < ApplicationController
   around_action :set_time_zone
+  before_action :set_global_create_context, only: %i[new]
```
And it's set this way

aaron-contreras committed:
Refactor/add spec coverage

aaron-contreras committed:
Remove dead code

Thanks for the catch Attila!

aaron-contreras commented:
Converting back to a draft as I'll be using Turbo to fetch the participant list when selecting a project from the dropdown

aaron-contreras committed:
Spice things up with Hotwire ;)

Adds a `turbo-frame` elemenet that dynamically fetches the participants section
upon a change to the selected project from the dropdown.

ulferts commented on modules/meeting/app/views/meetings/_form.html.erb:
```diff
@@ -86,48 +103,12 @@ See COPYRIGHT and LICENSE files for more details.
+  <%= turbo_frame_tag "participants_section",
+                      src: new_form_refresh_url,
+                      data: { refresh_on_form_changes_target: 'turboFrame' },
+                      loading: :lazy do %>
```
I don't think the `lazy` loading adds any value here as the element should be visible right from the beginning.

For the global creation case (selecting a project first) we wouldn't even need the `src` property to be set from the beginning as that will result in a useless request being issued where the response is always empty (except for the `turbo-frame`-tag.

In the project creation case no project is selected so the participant section will always stay empty. In there, the url should have the project included right from the start as that will be static anyway.

ulferts commented on modules/meeting/app/views/meetings/_form.html.erb:
```diff
@@ -35,6 +35,23 @@ See COPYRIGHT and LICENSE files for more details.
+                     controller: 'select-field-with-prompt',
+                     action: 'change->select-field-with-prompt#togglePromptStyling ' \
+                              'change->refresh-on-form-changes#triggerTurboReload'
+                   } %>
```
There is an angular powered autocompleter that sports all the niceties of a typeahead and also supports hierarchies:

<img width="589" alt="image" src="https://github.com/opf/openproject/assets/617519/7a644c4b-7740-4667-ad20-680b71d25b17">

It is e.g. used on the [page for bulk moving work packages](https://github.com/opf/openproject/blob/dev/app/views/work_packages/moves/new.html.erb#L65-L77). It should also be used here.

Doing this will require sending the correct filters for the all projects, in which the user is allowed to create meetings. Please see the [page for bulk moving work packages (the `apiFilters` property)](https://github.com/opf/openproject/blob/dev/app/views/work_packages/moves/new.html.erb#L67) as an example again. In order for that to work, the [`create_meetings` permission](https://github.com/opf/openproject/blob/dev/modules/meeting/lib/open_project/meeting/engine.rb#L43) needs to have `contract_actions` added, see [project create](https://github.com/opf/openproject/blob/dev/config/initializers/permissions.rb#L36) for an example.

ulferts commented on modules/meeting/app/controllers/meetings_controller.rb:
```diff
@@ -123,14 +128,24 @@ def set_time_zone(&)
+      @meeting.attributes = @converted_params
+    end
     @meeting.project = @project
     @meeting.author = User.current
```
Not necessary for this PR but it would make sense to move this functionality into a `Meetings::SetAttributesService`.

ulferts commented on modules/meeting/app/controllers/meetings_controller.rb:
```diff
@@ -123,14 +128,24 @@ def set_time_zone(&)
-    @project = Project.find(params[:project_id])
+    @project = Project.find(project_id)
     authorize
   rescue ActiveRecord::RecordNotFound
```
Please check whether this method overwrite over what is already implemented in [the `ApplicationController`](https://github.com/opf/openproject/blob/dev/app/controllers/application_controller.rb#L284-L288) is actually necessary.

aaron-contreras replied:
It is no longer necessary by using the `op-project-autocompleter` as I can have `project_id` be a non-nested param within the `meeting` key.

ulferts commented on modules/meeting/spec/features/meetings_new_spec.rb:
```diff
@@ -50,72 +50,150 @@
+
+          show_page.expect_date_time "03/28/2013 01:30 PM - 03:00 PM"
+        end
+      end
```
I don't understand why those passed on the CI. They fail for me locally and I think they do so rightfully as the `participants_section` does not show up on the project scoped new form.

aaron-contreras replied:
Now we know :)

ulferts reviewed (changes requested):
Hi @aaron-contreras please find my comments attached to the code. 

The part I am absolutely confused about is that the feature specs for the project scoped creation of meetings do not report the missing participant section.

aaron-contreras committed:
Address Jens's PR feedback

Thanks :)

aaron-contreras committed:
Provide a customizable `placeholder` option for `op-project-autocompleter`

**NOTE**: `typeToSearchText` was not an accepted option by `ng-select` according to its
docs. Removing it as it's dead code and the `placeholder` property was already provided
and this is where the "Type to search" placeholder was actually being fetched from.

aaron-contreras committed:
Appease Rubocop!

aaron-contreras committed:
Fix specs that didn't require a project to be present

aaron-contreras commented:
As discussed, I've moved the PR back to a draft state in order to modify the form implementation to use turbo-streams instead of turbo-frames.

aaron-contreras committed:
Modify `turbo-frame` approach to use `turbo-stream`s

As suggested by Jonas, this is much cleaner and extensible of an approach.
The `src` attribute hack of the prior `turbo-frame` approach is not really
of my liking for this anymore. As a side-benefit, this is quite a reusable
Stimulus action now.

ulferts commented on modules/meeting/app/helpers/meetings_helper.rb:
```diff
@@ -72,4 +72,14 @@ def render_journal_details(journal, header_label = :label_updated_time_by, _mode
+    request.path == new_meeting_path
+  end
+
+  def options_for_project_selection
```
This can now be removed since the functionality is now provided by the filters passed into the autocompleter.

ulferts commented on modules/meeting/config/locales/en.yml:
```diff
@@ -99,6 +101,8 @@ en:
 
   project_module_meetings: "Meetings"
 
+  project_selection_placeholder: "Select project"
```
Can be removed since the autocompleter is implemented in Angular now.

ulferts commented on modules/meeting/app/controllers/meetings_controller.rb:
```diff
@@ -127,15 +128,6 @@ def build_meeting
-  rescue ActiveRecord::RecordNotFound
-    render_404
-  end
-
```
Always good to get rid of those controller specific overrides if they are not needed. 🙇

ulferts commented on modules/meeting/lib/open_project/meeting/engine.rb:
```diff
@@ -40,7 +40,10 @@ class Engine < ::Rails::Engine
                                    meeting_minutes: %i[history show diff]
-        permission :create_meetings, { meetings: %i[new create copy] }, require: :member
+        permission :create_meetings,
+                   { meetings: %i[new participants_section create copy] },
```
With the change to turbo-stream, there no longer is a `participants_section` method added to the controller so this addition can be reverted.

ulferts reviewed (changes requested):
The things I am complaining about are leftovers of the pervious attempt. Feel free to merge this PR yourself, @aaron-contreras after those have been removed since the overall change is good.

aaron-contreras committed:
Remove leftover code from a previous implementation

aaron-contreras committed:
Merge branch 'dev' into implementation/48114-create-meetings-outside-of-a-project
